lib/Dialect/FIRRTL/Transforms/WireDFT.cpp

```diff
@@ -130,54 +134,69 @@ void WireDFTPass::runOnOperation() {
-                            module.getArgument(i), anno.getFieldID());
-      enableModule = module;
-      return true;
-    });
-    if (error)
-      return signalPassFailure();
 
-    // Walk the module body looking for any operation marked as the DFT enable.
-    auto walkResult = module->walk([&](Operation *op) {
-      AnnotationSet::removeAnnotations(op, [&](Annotation anno) {
-        // If we have already encountered an error or this is not the right
-        // annotation kind, continue.
-        if (error || !anno.isClass(dftTestModeEnableAnnoClass))
-          return false;
-        if (enableSignal) {
+    auto handleAnnotation = [&](Value value, Annotation anno,
+                                StringRef annoName, Value &signal,
+                                auto &signalModule) {
+      if (anno.isClass(annoName)) {
+        if (signal) {
           auto diag =
-              op->emitError("more than one thing marked as a DFT enable");
-          diag.attachNote(enableSignal.getLoc()) << "first thing defined here";
+              emitError(value.getLoc(), "more than one thing marked as ")
+              << annoName;
+          diag.attachNote(signal.getLoc()) << "first thing defined here";
           error = true;
           return false;
         }
+
         // Grab the enable value and remove the annotation.
-        enableSignal = getValueByFieldID(
-            ImplicitLocOpBuilder::atBlockEnd(op->getLoc(), op->getBlock()),
-            op->getResult(0), anno.getFieldID());
-        enableModule = module;
+        auto builder = ImplicitLocOpBuilder::atBlockEnd(value.getLoc(),
+                                                        value.getParentBlock());
+        builder.setInsertionPointAfterValue(value);
+        signal = getValueByFieldID(builder, value, anno.getFieldID());
+        signalModule = module;
         return true;
+      }
+      return false;
+    };
+
+    // See if this module has any port marked as the DFT enable.
+    AnnotationSet::removePortAnnotations(
+        module, [&](unsigned i, Annotation anno) {
+          if (!error && handleAnnotation(module.getArgument(i), anno,
+                                         dftTestModeEnableAnnoClass,
+                                         enableSignal, enableModule))
+            return true;
+
+          if (!error &&
+              handleAnnotation(module.getArgument(i), anno,
+                               dftClockDividerBypassAnnoClass,
+                               clockDivBypassSignal, clockDivBypassModule))
+            return true;
+
+          return false;
+        });
+    if (error)
+      return signalPassFailure();
+
+    // Walk the module body looking for any operation marked as the
+    // DFT enable.
+    auto walkResult = module->walk([&](Operation *op) {
+      AnnotationSet::removeAnnotations(op, [&](Annotation anno) {
+        if (op->getNumResults() == 0)
+          return false;
```

Comment:
Would slightly prefer keeping the early error check, the code is looking a little wonky below
```
if (error || op->getNumResults() == 0)
  return false;
```

Reply:
Refactored a bit, hopefully clearer?